apps/accounts/views.py

```diff
@@ -72,72 +99,100 @@ def login_user(request, user, identity):
     login(request, user)
 
 
-def with_user(fn):
-    @functools.wraps(fn)
-    def inner(self, request):
-        data = request.GET if request.method == 'GET' else request.DATA
-        state_parts = data.get('state', '').split(':', 1)
-        state = state_parts[0]
-        if 'code' not in data:
-            log.info('No code provided.')
-            return Response({'error': 'No code provided.'}, status=422)
-        elif (not request.session.get('fxa_state') or
-                request.session['fxa_state'] != state):
-            log.info('State mismatch. URL: {url} Session: {session}'.format(
-                url=data.get('state'),
-                session=request.session.get('fxa_state'),
-            ))
-            return Response({'error': 'State mismatch.'}, status=400)
+def render_error(request, error, next_path=None, format=None):
+    if format == 'json':
+        status = ERROR_STATUSES[error]
+        return Response({'error': error}, status=status)
+    else:
+        if not is_safe_url(next_path):
+            next_path = None
+        return HttpResponseRedirect(
+            urlparams(reverse('users.login'), error=error, to=next_path))
+
 
+def parse_next_path(state_parts):
+    next_path = None
+    if len(state_parts) == 2:
+        # The = signs will be stripped off so we need to add them back
+        # but it only cares if there are too few so add 4 of them.
+        encoded_path = state_parts[1] + '===='
```

Comment:
It just normalises the data into something we know will transfer well. We could probably URL encode it instead but it's going into a URL and that seems weird to me. I also think FxA was decoding `+`s or something so that could mess things up.

As for the `====` being needed:

```py
>>> base64.urlsafe_b64encode('quux')
'cXV1eA=='
>>> base64.urlsafe_b64decode('cXV1eA')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/base64.py", line 112, in urlsafe_b64decode
    return b64decode(s, '-_')
  File "/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/base64.py", line 76, in b64decode
    raise TypeError(msg)
TypeError: Incorrect padding
>>> base64.urlsafe_b64decode('cXV1eA====')
'quux'